File: Difficulty: Medium/Minimum Increment or Double Operations to Convert/minimum-increment-or-double-operations-to-convert.cpp
class Solution {
  public:
    int countMinOperations(vector<int>& arr) {
        int totalSetBits = 0;
        int maxBitPosition = 0; // Tracks maximum double operations needed

        for (int x : arr) {
            int bitPos = 0;

            while (x > 0) {
                // If LSB is set, it represents a +1 operation
                if (x & 1) {
                    totalSetBits++;
                }
                x >>= 1;
                bitPos++;
            }

            // Number of double operations for value 'x' is (bitPos - 1)
            if (bitPos > 0) {
                maxBitPosition = max(maxBitPosition, bitPos - 1);
            }
        }

        return totalSetBits + maxBitPosition;
    }
};

File: Difficulty: Medium/Minimum Increment or Double Operations to Convert/minimum-increment-or-double-operations-to-convert.java
class Solution {
    public int countMinOperations(int arr[]) {
        int ans = 0, maxDouble = 0;
        for(int i=0;i<arr.length;i++){
            int x = arr[i];
            int incrementOperation = 0;
            int doubleOperation = 0;
            while(x!=0){
                if(x%2==0){
                    doubleOperation++;
                    x /= 2;
                }else{
                    incrementOperation++;
                    x--;
                }
            }
            // Increment operation of an element is independent 
            // and cannot be clubbed with group operation
            ans += incrementOperation;
            // Double operation is common for all elements
            // So we find the max double operation per element
            maxDouble = Math.max(maxDouble, doubleOperation);
        }
        
        return ans + maxDouble;
    }
}